autoload/neomake/makers/ft/haskell.vim

```diff
@@ -1,55 +1,61 @@
 unlet! s:makers
 
-function! s:MakerAvailable(command) abort
-    " stack may be able to find a maker binary that's not on the normal path
-    " so check for that first
-    if executable('stack')
-        " run the maker command using stack to see whether stack can find it
-        " use the help flag to run the maker command without doing anything
-        if system('stack exec -- ' . a:command . ' --help > /dev/null 2>&1; echo $?') == 0
-            return 1
-        else " if stack cannot find the maker command, its not available anywhere
-            return 0
-        endif
-    elseif executable(a:command) " stack isn't available, so check for the maker binary directly
-        return 1
-    else
-        return 0
-    endif
-endfunction
-
 function! neomake#makers#ft#haskell#EnabledMakers() abort
-    " cache whether each maker is available, to avoid lots of (UI blocking) system calls...the user must restart vim if a maker's availability changes
-    if !exists('s:makers')
+    " cache whether each maker is available, to avoid lots of (UI blocking)
+    " system calls...unless we can check async, in which we do that instead,
+    " since it allows us to change directories from within vim and reasses
+    " maker availability without restarting vim
+    if !exists('s:makers') || has('nvim')
         let commands = ['ghc-mod', 'hdevtools', 'hlint', 'liquid']
         let s:makers = []
+        let s:jobs = []
         for command in commands
-            if s:MakerAvailable(command)
+            " stack may be able to find a maker binary that's not on the normal
+            " path so check for that first
+            if executable('stack')
+                " run the maker command using stack to see whether stack can
+                " find it use the help flag to run the maker command without
+                " doing anything
+                let stack_command = [
+                      \   'stack'
+                      \ , 'exec'
+                      \ , '--'
+                      \ , command
+                      \ , '--help'
+                      \ ]
+                if has('nvim')
```

Comment:
It would only be something in that regard, not exactly this.
For `jobwait` there would be a separate function then, and it looks like for Vim that would need to wait manually, using `job_status()` and `sleep` then.
It was just a suggestion to make it work also in Vim, and an opportunity for some refactoring / simplification.  I can see that this might be asked too much though definitely.. ;)